pydicom/filereader.py

```diff
@@ -501,6 +502,15 @@ def _not_group_0002(tag, VR, length):
                         "bytes)."
                         .format(file_meta.FileMetaInformationGroupLength,
                                 length_file_meta))
+
+    # Test the file meta for correct interpretation: if it fails, retry
+    #   loading the file meta with an implicit VR (issue #503)
+    try:
+        file_meta.get("TransferSyntaxUID")
+    except NotImplementedError:
+        fp.seek(start_file_meta)
+        file_meta = _read_file_meta_info(fp, is_implicit_VR=True)
```

Comment:
One problem I see here is that this means _read_file_meta_info calls itself.  Not a problem in itself, but what if the implicit VR case also doesn't work?  Then I think we have an infinite recursion.  I think the `try` block has to be moved up just after the `read_dataset` call, and in the catch just repeat that `read_dataset` line but with implicit VR.  Then the new optional argument to `_read_file_meta_info` is not necessary.

Reply:
And yes, we should have unit tests.  And given my review comment above, maybe we also need to add a test case where both implicit and explicit are wrong (perhaps the file is not even dicom), and test that the error is still raised.

Reply:
Good point. I decided to put the code in _read_file_meta_info because I needed the rollback to the start of the file meta when the exception is raised, but you are right this can lead to infinite recursion. Moving the code to just after the _read_file_meta_info call in read_partial seems better. And I'll look into unit tests!

Reply:
Good point, I didn't think of the infinite recursion. My mistake! I'll fix it and add unit tests. 

As for the case when neither implicit nor explicit VR works, do you have any suggestions for other options? Of the other transfer syntaxes, the only ones that I think are plausible for a file meta dataset are the big-endian options (please correct me if I'm wrong!). However, adding these would be tricky: a big-endian file meta dataset would be interpreted first as little-endian, but then any group 0002 tags are interpreted as 0200 tags, so the _read_file_meta will read an empty meta dataset and be happy with it. That could perhaps be solved, but it seems to me that would make it too complicated. What do you think?

For non-DICOM files the file meta is not really relevant I think: is such cases, pydicom will not read any file meta, and proceeds as it always has: the bytes of the file are interpreted as if it they represent valid dicom data, leading to either a KeyError (Unknown DICOM tag) or a dataset consisting of a single long private tag. What’s your view on this?